File: scripts/validate-jekyll-nav.py
#!/usr/bin/env python3
"""
Jekyll Navigation Validator for Just the Docs Theme

Validates that all .md files in the Jekyll site follow the navigation scheme:
- Every page has required front matter (title, nav_order)
- Parent references are valid (point to existing pages)
- Pages with has_children actually have children
- No use of grand_parent (not supported)
- No duplicate titles (parent matching is by title)
- All pages are reachable in the navigation hierarchy

Exit codes:
  0 - All validations passed
  1 - Validation errors found
  2 - Script error (e.g., missing dependencies)
"""

import os
import re
import sys
from pathlib import Path
from typing import Dict, List, Optional, Set, Tuple, Any

# Directories/files to exclude from validation
EXCLUDE_PATTERNS = [
    '.claude',
    '_site',
    '_layouts',
    '_includes',
    '_sass',
    'node_modules',
    '.git',
    'README.md',
    'CLAUDE.md',
]


def parse_front_matter(content: str) -> Optional[Dict[str, Any]]:
    """
    Parse YAML front matter from markdown content.
    Returns None if no valid front matter found.
    """
    # Match YAML front matter between --- delimiters
    match = re.match(r'^---\s*\n(.*?)\n---\s*\n', content, re.DOTALL)
    if not match:
        return None

    yaml_content = match.group(1)
    front_matter: Dict[str, Any] = {}

    # Simple YAML parsing (handles the cases we need)
    for line in yaml_content.split('\n'):
        line = line.strip()
        if not line or line.startswith('#'):
            continue

        if ':' in line:
            key, value = line.split(':', 1)
            key = key.strip()
            value = value.strip()

            # Handle boolean values
            if value.lower() == 'true':
                front_matter[key] = True
            elif value.lower() == 'false':
                front_matter[key] = False
            # Handle numeric values
            elif value.isdigit():
                front_matter[key] = int(value)
            # Handle quoted strings
            elif (value.startswith('"') and value.endswith('"')) or \
                 (value.startswith("'") and value.endswith("'")):
                front_matter[key] = value[1:-1]
            else:
                front_matter[key] = value

    return front_matter


def find_jekyll_files(root_dir: Path) -> List[Path]:
    """
    Find all .md files that are part of the Jekyll site.
    Excludes files in directories matching EXCLUDE_PATTERNS.
    """
    jekyll_files = []

    for md_file in root_dir.rglob('*.md'):
        # Get relative path for checking exclusions
        rel_path = md_file.relative_to(root_dir)
        rel_path_str = str(rel_path)

        # Check if file should be excluded
        should_exclude = False
        for pattern in EXCLUDE_PATTERNS:
            if rel_path_str.startswith(pattern) or rel_path.name == pattern:
                should_exclude = True
                break

        if not should_exclude:
            jekyll_files.append(md_file)

    return sorted(jekyll_files)


class ValidationError:
    """Represents a validation error."""

    def __init__(self, file_path: Path, message: str, severity: str = 'error'):
        self.file_path = file_path
        self.message = message
        self.severity = severity  # 'error' or 'warning'

    def __str__(self) -> str:
        return f"[{self.severity.upper()}] {self.file_path}: {self.message}"


class JekyllNavValidator:
    """Validates Jekyll navigation structure."""

    def __init__(self, root_dir: Path):
        self.root_dir = root_dir
        self.pages: Dict[Path, Dict[str, Any]] = {}
        self.titles: Dict[str, Path] = {}  # title -> file path
        self.errors: List[ValidationError] = []

    def load_pages(self) -> None:
        """Load and parse all Jekyll pages."""
        jekyll_files = find_jekyll_files(self.root_dir)

        for file_path in jekyll_files:
            try:
                content = file_path.read_text(encoding='utf-8')
                front_matter = parse_front_matter(content)

                if front_matter is not None:
                    self.pages[file_path] = front_matter
            except Exception as e:
                self.errors.append(ValidationError(
                    file_path,
                    f"Failed to read file: {e}"
                ))

    def validate_required_fields(self) -> None:
        """Validate that all pages have required front matter fields."""
        for file_path, front_matter in self.pages.items():
            # Check for title
            if 'title' not in front_matter:
                self.errors.append(ValidationError(
                    file_path,
                    "Missing required 'title' in front matter"
                ))

            # Check for nav_order
            if 'nav_order' not in front_matter:
                self.errors.append(ValidationError(
                    file_path,
                    "Missing required 'nav_order' in front matter"
                ))

    def build_title_index(self) -> None:
        """Build an index of titles to file paths and check for duplicates."""
        for file_path, front_matter in self.pages.items():
            title = front_matter.get('title')
            if title:
                if title in self.titles:
                    self.errors.append(ValidationError(
                        file_path,
                        f"Duplicate title '{title}' (also in {self.titles[title].relative_to(self.root_dir)})"
                    ))
                else:
                    self.titles[title] = file_path

    def validate_no_grand_parent(self) -> None:
        """Validate that no pages use the unsupported grand_parent property."""
        for file_path, front_matter in self.pages.items():
            if 'grand_parent' in front_matter:
                self.errors.append(ValidationError(
                    file_path,
                    "Using 'grand_parent' which is not supported - use 'parent' pointing to the child page's title instead"
                ))

    def validate_parent_references(self) -> None:
        """Validate that all parent references point to existing pages."""
        for file_path, front_matter in self.pages.items():
            parent = front_matter.get('parent')
            if parent:
                if parent not in self.titles:
                    self.errors.append(ValidationError(
                        file_path,
                        f"Parent '{parent}' does not match any page title"
                    ))
                else:
                    # Check that parent has has_children: true
                    parent_path = self.titles[parent]
                    parent_front_matter = self.pages[parent_path]
                    if not parent_front_matter.get('has_children'):
                        self.errors.append(ValidationError(
                            parent_path,
                            f"Page is referenced as parent by '{front_matter.get('title', file_path.name)}' but lacks 'has_children: true'"
                        ))

    def validate_has_children(self) -> None:
        """Validate that pages with has_children actually have children."""
        # Build set of all parent titles
        parent_titles: Set[str] = set()
        for front_matter in self.pages.values():
            parent = front_matter.get('parent')
            if parent:
                parent_titles.add(parent)

        # Check pages with has_children
        for file_path, front_matter in self.pages.items():
            if front_matter.get('has_children'):
                title = front_matter.get('title')
                if title and title not in parent_titles:
                    self.errors.append(ValidationError(
                        file_path,
                        f"Page has 'has_children: true' but no pages reference it as parent",
                        severity='warning'
                    ))

    def validate_reachability(self) -> None:
        """Validate that all pages are reachable in the navigation hierarchy."""
        for file_path, front_matter in self.pages.items():
            title = front_matter.get('title')
            if not title:
                continue

            # If page has no parent, it must have nav_order to be top-level
            if 'parent' not in front_matter:
                if 'nav_order' not in front_matter:
                    self.errors.append(ValidationError(
                        file_path,
                        "Page has no parent and no nav_order - will not appear in navigation"
                    ))
            else:
                # Trace parent chain to ensure it reaches a top-level page
                visited = {title}
                current_parent = front_matter.get('parent')

                while current_parent:
                    if current_parent in visited:
                        self.errors.append(ValidationError(
                            file_path,
                            f"Circular parent reference detected involving '{current_parent}'"
                        ))
                        break

                    if current_parent not in self.titles:
                        # Already reported as invalid parent
                        break

                    visited.add(current_parent)
                    parent_path = self.titles[current_parent]
                    parent_front_matter = self.pages[parent_path]
                    current_parent = parent_front_matter.get('parent')

    def validate_nav_order_uniqueness(self) -> None:
        """Validate that nav_order is unique among siblings."""
        # Group pages by their parent
        siblings: Dict[Optional[str], List[Tuple[Path, Dict[str, Any]]]] = {}

        for file_path, front_matter in self.pages.items():
            parent = front_matter.get('parent')
            if parent not in siblings:
                siblings[parent] = []
            siblings[parent].append((file_path, front_matter))

        # Check for duplicate nav_order within each group
        for parent, pages in siblings.items():
            nav_orders: Dict[int, Path] = {}
            for file_path, front_matter in pages:
                nav_order = front_matter.get('nav_order')
                if nav_order is not None:
                    if nav_order in nav_orders:
                        context = f"under parent '{parent}'" if parent else "at top level"
                        self.errors.append(ValidationError(
                            file_path,
                            f"Duplicate nav_order {nav_order} {context} (also in {nav_orders[nav_order].relative_to(self.root_dir)})",
                            severity='warning'
                        ))
                    else:
                        nav_orders[nav_order] = file_path

    def validate(self) -> bool:
        """
        Run all validations and return True if no errors found.
        """
        self.load_pages()

        if not self.pages:
            print("Warning: No Jekyll pages found to validate")
            return True

        # Run validations in order
        self.validate_required_fields()
        self.build_title_index()
        self.validate_no_grand_parent()
        self.validate_parent_references()
        self.validate_has_children()
        self.validate_reachability()
        self.validate_nav_order_uniqueness()

        return len([e for e in self.errors if e.severity == 'error']) == 0

    def print_report(self) -> None:
        """Print validation report."""
        if not self.errors:
            print(f"All {len(self.pages)} Jekyll pages passed navigation validation")
            return

        errors = [e for e in self.errors if e.severity == 'error']
        warnings = [e for e in self.errors if e.severity == 'warning']

        print(f"\nJekyll Navigation Validation Report")
        print("=" * 50)
        print(f"Pages scanned: {len(self.pages)}")
        print(f"Errors: {len(errors)}")
        print(f"Warnings: {len(warnings)}")
        print()

        if errors:
            print("ERRORS:")
            print("-" * 50)
            for error in errors:
                rel_path = error.file_path.relative_to(self.root_dir)
                print(f"  {rel_path}")
                print(f"    {error.message}")
            print()

        if warnings:
            print("WARNINGS:")
            print("-" * 50)
            for warning in warnings:
                rel_path = warning.file_path.relative_to(self.root_dir)
                print(f"  {rel_path}")
                print(f"    {warning.message}")
            print()


def main() -> int:
    """Main entry point."""
    # Determine root directory
    if len(sys.argv) > 1:
        root_dir = Path(sys.argv[1]).resolve()
    else:
        # Use current working directory or find project root
        root_dir = Path.cwd()

        # Try to find project root by looking for _config.yml or index.md
        check_dir = root_dir
        while check_dir != check_dir.parent:
            if (check_dir / '_config.yml').exists() or \
               (check_dir / 'index.md').exists():
                root_dir = check_dir
                break
            check_dir = check_dir.parent

    if not root_dir.exists():
        print(f"Error: Directory not found: {root_dir}", file=sys.stderr)
        return 2

    print(f"Validating Jekyll navigation in: {root_dir}")

    validator = JekyllNavValidator(root_dir)
    success = validator.validate()
    validator.print_report()

    return 0 if success else 1


if __name__ == '__main__':
    sys.exit(main())

File: scripts/validate-jekyll-nav.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Jekyll Navigation Validator for Just the Docs Theme
#
# Validates that all .md files in the Jekyll site follow the navigation scheme:
# - Every page has required front matter (title, nav_order)
# - Parent references are valid (point to existing pages)
# - Pages with has_children actually have children
# - No use of grand_parent (not supported)
# - No duplicate titles (parent matching is by title)
# - All pages are reachable in the navigation hierarchy
#
# Exit codes:
#   0 - All validations passed
#   1 - Validation errors found
#   2 - Script error

require 'yaml'
require 'pathname'

# Directories/files to exclude from validation
EXCLUDE_PATTERNS = %w[
  .claude
  _site
  _layouts
  _includes
  _sass
  node_modules
  .git
  README.md
  CLAUDE.md
].freeze

class ValidationError
  attr_reader :file_path, :message, :severity

  def initialize(file_path, message, severity: :error)
    @file_path = file_path
    @message = message
    @severity = severity
  end

  def to_s
    "[#{severity.to_s.upcase}] #{file_path}: #{message}"
  end
end

class JekyllNavValidator
  attr_reader :root_dir, :pages, :titles, :errors

  def initialize(root_dir)
    @root_dir = Pathname.new(root_dir).expand_path
    @pages = {}          # file_path => front_matter hash
    @titles = {}         # title => file_path
    @errors = []
  end

  def find_jekyll_files
    files = []
    Dir.glob(@root_dir.join('**/*.md')).each do |file|
      file_path = Pathname.new(file)
      rel_path = file_path.relative_path_from(@root_dir).to_s

      # Check if file should be excluded
      should_exclude = EXCLUDE_PATTERNS.any? do |pattern|
        rel_path.start_with?(pattern) || File.basename(rel_path) == pattern
      end

      files << file_path unless should_exclude
    end
    files.sort
  end

  def parse_front_matter(content)
    return nil unless content.start_with?('---')

    # Find the closing ---
    second_marker = content.index('---', 3)
    return nil unless second_marker

    yaml_content = content[3...second_marker].strip
    begin
      YAML.safe_load(yaml_content, permitted_classes: [Symbol])
    rescue Psych::SyntaxError => e
      nil
    end
  end

  def load_pages
    find_jekyll_files.each do |file_path|
      begin
        content = File.read(file_path, encoding: 'UTF-8')
        front_matter = parse_front_matter(content)

        if front_matter.is_a?(Hash)
          @pages[file_path] = front_matter
        end
      rescue => e
        @errors << ValidationError.new(
          file_path.relative_path_from(@root_dir),
          "Failed to read file: #{e.message}"
        )
      end
    end
  end

  def validate_required_fields
    @pages.each do |file_path, front_matter|
      rel_path = file_path.relative_path_from(@root_dir)

      unless front_matter['title']
        @errors << ValidationError.new(
          rel_path,
          "Missing required 'title' in front matter"
        )
      end

      unless front_matter['nav_order']
        @errors << ValidationError.new(
          rel_path,
          "Missing required 'nav_order' in front matter"
        )
      end
    end
  end

  def build_title_index
    @pages.each do |file_path, front_matter|
      title = front_matter['title']
      next unless title

      rel_path = file_path.relative_path_from(@root_dir)

      if @titles[title]
        @errors << ValidationError.new(
          rel_path,
          "Duplicate title '#{title}' (also in #{@titles[title]})"
        )
      else
        @titles[title] = rel_path
      end
    end
  end

  def validate_no_grand_parent
    @pages.each do |file_path, front_matter|
      next unless front_matter['grand_parent']

      rel_path = file_path.relative_path_from(@root_dir)
      @errors << ValidationError.new(
        rel_path,
        "Using 'grand_parent' which is not supported - use 'parent' pointing to the child page's title instead"
      )
    end
  end

  def validate_parent_references
    @pages.each do |file_path, front_matter|
      parent = front_matter['parent']
      next unless parent

      rel_path = file_path.relative_path_from(@root_dir)

      unless @titles[parent]
        @errors << ValidationError.new(
          rel_path,
          "Parent '#{parent}' does not match any page title"
        )
        next
      end

      # Find parent's front matter and check has_children
      parent_file = @pages.find { |_, fm| fm['title'] == parent }
      if parent_file
        parent_front_matter = parent_file[1]
        unless parent_front_matter['has_children']
          @errors << ValidationError.new(
            @titles[parent],
            "Page is referenced as parent by '#{front_matter['title'] || rel_path}' but lacks 'has_children: true'"
          )
        end
      end
    end
  end

  def validate_has_children
    # Build set of all parent titles
    parent_titles = Set.new
    @pages.each_value do |front_matter|
      parent = front_matter['parent']
      parent_titles << parent if parent
    end

    # Check pages with has_children
    @pages.each do |file_path, front_matter|
      next unless front_matter['has_children']

      title = front_matter['title']
      next unless title

      unless parent_titles.include?(title)
        rel_path = file_path.relative_path_from(@root_dir)
        @errors << ValidationError.new(
          rel_path,
          "Page has 'has_children: true' but no pages reference it as parent",
          severity: :warning
        )
      end
    end
  end

  def validate_reachability
    @pages.each do |file_path, front_matter|
      title = front_matter['title']
      next unless title

      rel_path = file_path.relative_path_from(@root_dir)

      # If page has no parent, it must have nav_order to be top-level
      unless front_matter['parent']
        unless front_matter['nav_order']
          @errors << ValidationError.new(
            rel_path,
            "Page has no parent and no nav_order - will not appear in navigation"
          )
        end
        next
      end

      # Trace parent chain to ensure it reaches a top-level page
      visited = Set.new([title])
      current_parent = front_matter['parent']

      while current_parent
        if visited.include?(current_parent)
          @errors << ValidationError.new(
            rel_path,
            "Circular parent reference detected involving '#{current_parent}'"
          )
          break
        end

        unless @titles[current_parent]
          # Already reported as invalid parent
          break
        end

        visited << current_parent

        # Find parent's front matter
        parent_file = @pages.find { |_, fm| fm['title'] == current_parent }
        break unless parent_file

        current_parent = parent_file[1]['parent']
      end
    end
  end

  def validate_nav_order_uniqueness
    # Group pages by their parent
    siblings = Hash.new { |h, k| h[k] = [] }

    @pages.each do |file_path, front_matter|
      parent = front_matter['parent']
      siblings[parent] << [file_path, front_matter]
    end

    # Check for duplicate nav_order within each group
    siblings.each do |parent, pages_list|
      nav_orders = {}
      pages_list.each do |file_path, front_matter|
        nav_order = front_matter['nav_order']
        next unless nav_order

        rel_path = file_path.relative_path_from(@root_dir)

        if nav_orders[nav_order]
          context = parent ? "under parent '#{parent}'" : "at top level"
          @errors << ValidationError.new(
            rel_path,
            "Duplicate nav_order #{nav_order} #{context} (also in #{nav_orders[nav_order]})",
            severity: :warning
          )
        else
          nav_orders[nav_order] = rel_path
        end
      end
    end
  end

  def validate
    load_pages

    if @pages.empty?
      puts "Warning: No Jekyll pages found to validate"
      return true
    end

    # Run validations in order
    validate_required_fields
    build_title_index
    validate_no_grand_parent
    validate_parent_references
    validate_has_children
    validate_reachability
    validate_nav_order_uniqueness

    @errors.none? { |e| e.severity == :error }
  end

  def print_report
    if @errors.empty?
      puts "All #{@pages.size} Jekyll pages passed navigation validation"
      return
    end

    error_list = @errors.select { |e| e.severity == :error }
    warning_list = @errors.select { |e| e.severity == :warning }

    puts
    puts "Jekyll Navigation Validation Report"
    puts "=" * 50
    puts "Pages scanned: #{@pages.size}"
    puts "Errors: #{error_list.size}"
    puts "Warnings: #{warning_list.size}"
    puts

    unless error_list.empty?
      puts "ERRORS:"
      puts "-" * 50
      error_list.each do |error|
        puts "  #{error.file_path}"
        puts "    #{error.message}"
      end
      puts
    end

    unless warning_list.empty?
      puts "WARNINGS:"
      puts "-" * 50
      warning_list.each do |warning|
        puts "  #{warning.file_path}"
        puts "    #{warning.message}"
      end
      puts
    end
  end
end

def main
  # Determine root directory
  root_dir = if ARGV[0]
               File.expand_path(ARGV[0])
             else
               # Use current working directory or find project root
               dir = Dir.pwd
               while dir != File.dirname(dir)
                 if File.exist?(File.join(dir, '_config.yml')) ||
                    File.exist?(File.join(dir, 'index.md'))
                   break
                 end
                 dir = File.dirname(dir)
               end
               dir
             end

  unless File.directory?(root_dir)
    warn "Error: Directory not found: #{root_dir}"
    exit 2
  end

  puts "Validating Jekyll navigation in: #{root_dir}"

  validator = JekyllNavValidator.new(root_dir)
  success = validator.validate
  validator.print_report

  exit(success ? 0 : 1)
end

main if __FILE__ == $PROGRAM_NAME
